### PR TITLE
Create and activate product attributes lookup table in data migration

### LINKED ISSUE #30564: Convert the product attributes lookup table creation process to a data migration
[The new product filtering by attributes using a lookup table feature](https://developer.woocommerce.com/2021/08/12/brace-yourselves-new-product-filtering-by-attributes-is-coming/) is initially aimed at experienced users for testing. As such, the table needs to be created manually via a dedicated tool in the Status - Tools page.

The second step towards enabling the feature for everybody (after https://github.com/woocommerce/woocommerce/issues/30563) would be to create and fill the lookup table automatically in a data migration. The manual creation and filling currently in place is already done via scheduled actions, so most of the infrastructure needed should be already in place.

Note that while the tool to delete the table should be removed, the tool to regenerate the table contents could be left in place.

### LINKED ISSUE #30565: Enable the product attribute lookup table usage by default
[The new product filtering by attributes using a lookup table feature](https://developer.woocommerce.com/2021/08/12/brace-yourselves-new-product-filtering-by-attributes-is-coming/) is initially aimed at experienced users for testing. As such, the table usage is disabled by default, and needs to be enabled manually via a dedicated setting.

The last step towards enabling the feature for everybody (after https://github.com/woocommerce/woocommerce/issues/30563 and https://github.com/woocommerce/woocommerce/issues/30564) would be to enable the feature by default on new WooCommerce installs. The setting would still be left in place so that it can be disabled.

### PR DESCRIPTION
**Pull request by Konamiman**
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Until now the product attributes lookup table had to be created and its usage had to be activated manually, via the tools page. This pull request does the following changes:

1. Remove the tools to create and delete the table (but keep the tool to regenerate the table data for one or for all the products).

2. Add a data migration that triggers the table creation and filling (unless the table already existed, in this case nothing happens).

3. After the migration finishes, activate the table usage for product filtering (site admin can still disable it via _Settings - Products - Advanced_); again, if the table already existed, nothing happens (the setting is unchanged). The setting will be left disabled if the regeneration is abored manually (see below).

Additionally, the following has been added in order to help in the migration process:

1. A new filter, `woocommerce_attribute_lookup_regeneration_step_size`, that allows to alter the batch size for the process (the number of products that will be processed each time that Action Scheduler runs one migration step). This filter is run on every step, thus it can be used to alter the batch size even if the regeneration process has been already started.

2. If the regeneration process is aborted manually (by deleting the `woocommerce_attribute_lookup_regeneration_in_progress` option), the _Products - Advanced_ settings section will still allow to enable the lookup table usage, but will display a warning discouraging to do so because the lookup table will be in an inconsistent state. The warning is shown when the `woocommerce_attribute_lookup_regeneration_aborted` option exists, and the option is deleted when the table regeneration is triggered again via the tools page.

image

Closes #30564.
Closes #30565.

### How to test the changes in this Pull Request:

1. If you had tested the product attributes lookup table mechanism already and the table still exists in your db, delete it from _Status - Tools - Delete the product attributes lookup table_ (you need to do this from `trunk`).

2. Switch to the branch of this pull request and go to _Status - Tools_. You shouldn't see any entry related to the product attributes lookup table.

3. Trigger the WooCommerce db update from the update available warning in the admin area.

3. After the update finishes you'll see that the old tools page entry "Regenerate the product attributes lookup table" is back. If you have a lot of products in your db the "Regenerate" button will be disabled and will show a counter of processed products, this means that the table regeneration is still in progress.

4. Once the table filling is complete the "Regenerate" product should appear as enabled. After that has happened go to _Settings - Products - Advanced_ and verify that "Enable table usage" is set.

5. Verify that the feature still works as expected (enabling and disabling the table usage via settings, regenerating the table from tools page for one product or for all products).

Now, to test the case of a db update when the table already exists:

6. Disable the table usage via settings.

7. Run the following in command line: `wp option set woocommerce_db_version 6.0.0`

8. Trigger the db update from the banner again, this time nothing related to the lookup table happens (the "Regenerate" tool has its button enabled right after the db update finishes).

9. Verify that the table usage is still disabled in settings.

To test the new filter:

10. Add the following somewhere (e.g. at the end of `woocommerce.php`):

```php
add_filter('woocommerce_attribute_lookup_regeneration_step_size', function() {
	return 5;
});
```

Verify that the regeneration happens now in increments of five products (hint: use `wp option get woocommerce_attribute_lookup_processed_count` to easily get the count of products already processed). If you change the value mid-regeneration you should see that the new value is used from the following regeneration step.

To test the new warning message:

11. Abort an already started regeneration by running `wp option delete woocommerce_attribute_lookup_regeneration_in_progress`.

12. Verify that the _Products - Advanced_ settings page now shows the warning (you may need to reload a couple of times to give Action Scheduler a chance to run one last time).

13. Start a new regeneration process, let it finish, and verify that the warning has disappeared.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Add - Data migration to create and activate the product attributes lookup table.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by barryhughes** [comment]
✍🏽 Hold on merging p1637741479224500-slack-C0E1AV8T0 (internal ref).

**Comment by barryhughes** [comment]
🔬 **Observation regarding `OR` filtering**

There seems to be a quirk when a filter is configured to use the `OR` query type. Setup:

- I created an attribute called _'complexity'._
- Several terms belonging to this attribute are applied to lots of different products.
- One term called _'custom'_ is applied to just a single product.

With lookup table usage disabled and if I select _'custom'_ from this filter, then the widget looks like this (matching my expectations):

01-no-table-usage

With lookup table usage _enabled,_ the widget will look like this:

02-with-table-usage

There are two problems here:

- Apart from the selected term, the only other terms I now see are those that happen to be applied to the same product(s) as the selected term. Since we are using `OR` filtering, though, I'd expect _all_ attribute terms to be listed (except perhaps for those that are not applied to any products).
- The counts are incorrect. For example, if I were to additionally select _'fiendish'_ I would in fact see 215 products (and if I go ahead and do that, the counts are corrected).

**Comment by barryhughes** [comment]
🔬 **Observation: regenerating lookup data for a single product**

Very minor but noting for completeness: if I use the _regenerate the product attributes lookup table_ tool **for a single product** then (as I understand it) the relevant entries are generated by the time the page reloads, but we still see a message reading:

> Product attributes lookup table data is regenerating

In this particular case, it might be better for it to read:

> Product attribute data for <Product XYZ> was successfully regenerated.

File: plugins/woocommerce/includes/class-wc-install.php [hunk 1]
```diff
@@ -169,6 +169,10 @@ class WC_Install {
 			'wc_update_600_migrate_rate_limit_options',
 			'wc_update_600_db_version',
 		),
+		'6.1.0' => array(
```
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-install.php]: > Scheduled for WooCommerce 6.2

Given this note, should we adjust to `'6.2.0'` (and if so also adjust `wc_update_610_db_version()`)?

**Review by barryhughes — COMMENTED** [review]
Left a couple of notes to check on, but looking good 👍🏽

File: plugins/woocommerce/includes/class-wc-install.php (hunk 1 above)
Comment (Konamiman) [reply plugins/woocommerce/includes/class-wc-install.php:172]: Yep, makes sense. Fixed.

Comment (Konamiman) [reply plugins/woocommerce/includes/class-wc-install.php]: Yep, makes sense. Fixed.

**Comment by Konamiman** [comment]
> Very minor but noting for completeness: if I use the regenerate the product attributes lookup table tool for a single product then (as I understand it) the relevant entries are generated by the time the page reloads, but we still see a message reading:

Be careful, reloading the tools page is tricky. If you just reload, the page URL will still have the `action` query string parameter set, thus you are effectively triggering the regeneration process again. You need to manually remove the parameter from the URL before reloading.

**Comment by Konamiman** [comment]
> There seems to be a quirk when a filter is configured to use the OR query type

Indeed, looks like it was "overfiltering" in this case. I think I have fixed it.

**Review by barryhughes — DISMISSED** [review]
LGTM!

- Not merging (per instructions)
- Checks are failing, but as far as I can determine the fails are unrelated

**Comment by hossamadel99** [comment]
Hello ,

Followed all steps but when trying to filter any attribute it takes about 20 seconds, the page refreshes in two times, one instantly without any filtration but after 20 seconds succeed.

Any guide please regarding the speed with ( attributes lookup table )

**Comment by Konamiman** [comment]
Hi @hossamadel99 , are you ensuring that the table regeneration has finished before attempting to actually load the products page?

**Comment by hossamadel99** [comment]
> Hi @hossamadel99 , are you ensuring that the table regeneration has finished before attempting to actually load the products page?

Yes & got (Regenerate) button

**Comment by Konamiman** [comment]
@hossamadel99 I'd need more data in order to troubleshoot this. Is there any testing site I can take a look at? Or alternatively, could you please send me a product dump? You can reach me at nestor.soriano@automattic.com

**Comment by Konamiman** [comment]
Just noting here for reference that @hossamadel99 did indeed contact me via email and confirmed me that after upgrading to WooCommerce 6.0 the performance issue disappeared. Thanks!

File: plugins/woocommerce/includes/wc-update-functions.php [hunk 2]
```diff
@@ -2335,3 +2337,28 @@ function wc_update_600_migrate_rate_limit_options() {
...
 	WC_Install::update_db_version( '6.0.0' );
 }
+
+/**
+ * Create the product attributes lookup table and initiate its filling,
+ * unless the table had been already created manually (via the tools page).
+ *
+ * @return false Always false, since the LookupDataStore class handles all the data filling process.
+ */
+function wc_create_product_attributes_lookup_table() {
```
Comment (jeffstieler) [inline plugins/woocommerce/includes/wc-update-functions.php]: Should we continue the pattern of including the version number in the update function name here?

File: plugins/woocommerce/woocommerce.php [hunk 3]
```diff
@@ -60,3 +60,7 @@ function wc_get_container() : \Psr\Container\ContainerInterface {
 
 // Global for backwards compatibility.
 $GLOBALS['woocommerce'] = WC();
+
+/*add_filter('woocommerce_attribute_lookup_regeneration_step_size', function() {
+	return 5;
+});*/
```
Comment (jeffstieler) [inline plugins/woocommerce/woocommerce.php]: Looks like this might have been included by mistake.

**Review by jeffstieler — COMMENTED** [review]
Saw some minor things.

**Comment by barryhughes** [comment]
A little unsure if this is a problem or not (or even if it has to be addressed right away, if we do see it as a problem). At any rate, I wanted to note that if I use this tool:

verify-base-database-tables

Then it reports success:

verify-base-database-tables-successful

However, the product attributes lookup table is not present/has not yet been created. *Should* it be covered by this tool, or will it remain a distinct thing?

File: plugins/woocommerce/src/Internal/ProductAttributesLookup/LookupDataStore.php [hunk 4]
```diff
@@ -648,4 +657,28 @@ public function set_regeneration_in_progress_flag() {
...
+		delete_option( 'woocommerce_attribute_lookup_regeneration_aborted' );
+	}
+
+	/**
+	 * Tells if the last lookup table regeneration process started was aborted
+	 * (via deleting the 'woocommerce_attribute_lookup_regeneration_in_progress' option).
+	 *
+	 * @return bool True if the last lookup table regeneration process was aborted.
+	 */
+	public function regeneration_was_aborted() {
```
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/ProductAttributesLookup/LookupDataStore.php]: Minor thing, but we could use return types here:

```suggestion
	public function regeneration_was_aborted(): bool {
```

**Review by barryhughes — DISMISSED** [review]
LGTM. 

I left a couple of notes, so am not merging so you can review those first of all; but if no action is needed and we're otherwise ready to merge please go ahead.

**Comment by Konamiman** [comment]
@barryhughes The attributes lookup table should probably be included in the list of core WooCommerce tables, indeed; but that's also the case for a bunch of other already existing WooCommerce tables. So fixing this would better be done in a separate project and pull request.

**Review by barryhughes — APPROVED** [review]
Looks good

**Comment by barryhughes** [comment]
Going ahead with merge (just noting this because the description still states it shouldn't be merged, but per recent-ish team discussion this work is good to go).

**Pull request merged**